powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/v4/kdf/Kdf.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * PowerAuth Crypto Library
+ * Copyright 2024 Wultra s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getlime.security.powerauth.crypto.lib.v4.kdf;
+
+import io.getlime.security.powerauth.crypto.lib.model.exception.GenericCryptoException;
+import io.getlime.security.powerauth.crypto.lib.util.ByteUtils;
+import io.getlime.security.powerauth.crypto.lib.util.KeyConvertor;
+import org.bouncycastle.crypto.macs.KMAC;
+import org.bouncycastle.crypto.params.KeyParameter;
+
+import javax.crypto.SecretKey;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Universal KDF based on KMAC-256 (Keccak).
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+public class Kdf {
+
+    private static final String CRYPTO4_KDF_CUSTOM_STRING = "PA4KDF";
+    private static final Integer KMAC_ALGORITHM_STRENGTH = 256;
+
+    private static final KeyConvertor KEY_CONVERTOR = new KeyConvertor();
+
+    /**
+     * Derive a secret key based on an input key, numeric key index, requested key size and optional context.
+     * @param key Secret key to be used for key derivation.
+     * @param index Key index (numeric).
+     * @param outLength Requested derived key size.
+     * @param context Optional context to use during key derivation.
+     * @return Derived secret key.
+     * @throws GenericCryptoException Thrown in case of any cryptography error.
+     */
+    public static SecretKey derive(SecretKey key, long index, int outLength, byte[] context) throws GenericCryptoException {
+        final byte[] indexBytes = ByteUtils.encodeLong(index);
+        final byte[] data;
+        if (context != null) {
+            data = ByteUtils.concat(indexBytes, ByteUtils.concatWithSizes(context));
```

Comment:
Not sure if this can lead to ambiguous values, i.e., if `indexBytes == 'AAAA' && context == 'BBBB'`, vs. `indexBytes == null && context == 'AAAABBBB'` - I recommend either adding a separator, or adding a differentiator, or keeping the same length with zero prefix, ...

Reply:
@hvge please comment

Reply:
`index` is required parameter with fixed size length - 8 bytes (it's BE encoded long integer). The `context` is optional parameter and if it's provided, then it's always separated by the length. For example:
- context not provided, then CTX is: `INDEX (8 bytes)`
- context is provided, then CTX is `INDEX (8 bytes) ||  CONTEXT.length (4 bytes) || CONTEXT`

I don't see how this could be ambiguous.